### PR TITLE
Modified API of sm2/3/4

### PR DESCRIPTION
**Pull request by Messjer**
Basically I have done the following changes:
* Modified the API of the 3 algorithms to use bytes;
* Created Aliases for type names for clearer meanings;
* Changed the tests and benchmarks to fit the new API;
* Added some documentations for the functions.

I also fixed a tiny bug in `is_on_curve` of `mod ecc_group`.

I did not changed the API of sm3 to produce bytes as hash value. Instead I madee `HashValue` as alias for `[u32;4]`. This is because I believe user normally does not need to manipulate with these blocks as byte arrays anyways.

### CODE REVIEW

File: src/sm2/mod.rs [hunk 1]
```diff
@@ -1,16 +1,58 @@
...
+pub type SecKey = U64x4;
+pub struct Signature {
+    pub r: U64x4,
+    pub s: U64x4,
+}
+
+/// Randomly sample secret key uniformly from [0,..n), where n is the order of the base point
+pub fn get_sec_key() -> SecKey {
+    // TODO replace with random
+    U64x4::new(0, 0, 0, 0)
```
Comment (mssun) [inline src/sm2/mod.rs]: Can you merge changes from master and replace this?

File: src/basic/helper.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+//! A helper module for commonly used internal functions
```
Comment (mssun) [inline src/basic/helper.rs]: We always use `util.rs` for utility functions in Rust.

File: src/sm4/mod.rs [hunk 3]
```diff
@@ -221,6 +247,49 @@ macro_rules! sm4_rnds {
     };
 }
 
+fn bytes_to_four_u32(b: &[u8; 16]) -> [u32; 4] {
```
Comment (mssun) [inline src/sm4/mod.rs]: Can we move this to `util.rs`?

File: src/sm4/mod.rs [hunk 4]
```diff
@@ -221,6 +247,49 @@ macro_rules! sm4_rnds {
...
+            + (u32::from(b[10]) << 8)
+            + u32::from(b[11]),
+        (u32::from(b[12]) << 24)
+            + (u32::from(b[13]) << 16)
+            + (u32::from(b[14]) << 8)
+            + u32::from(b[15]),
+    ]
+}
+
+fn four_u32_to_bytes(l: &[u32; 4]) -> [u8; 16] {
```
Comment (mssun) [inline src/sm4/mod.rs]: Can we move this to `util.rs`?

Comment (mssun) [reply src/sm4/mod.rs]: Can we come up better names for these:`bytes_to_four_u32`, `four_u32_to_bytes`? Anyway, I don't have better idea for now.

**Review by mssun — CHANGES_REQUESTED** [review]
Very good! Only need some minor changes. Thanks.

**Comment by mssun** [comment]
BTW, `clippy` is not happy. Can you fix it? https://travis-ci.com/yogcrypt/yogcrypt/jobs/139484762

**Comment by Messjer** [comment]
Thanks for the review. I fix these issues in 41c7e907e97c3adcccd0a323c43de00d15af7026.

**Comment by mssun** [comment]
Great, thanks.

**Review by mssun — APPROVED** [review]

**Comment by mssun** [comment]
Merged, please delete the remote branch if you don't need it.

**Pull request merged**